<a href="https://colab.research.google.com/github/JonMaycon/DataScience/blob/main/WS_CamaraVereadores_Cba.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Webscraping Projetos de Lei - Camara de Vereadores de Cuiabá - MT

In [1]:
import requests
from bs4 import BeautifulSoup

In [2]:
url = "https://legislativo.camaracuiaba.mt.gov.br/consulta-producao.aspx?tipo=103"

headers = {}

response  = requests.get(url, headers=headers)

soup = BeautifulSoup(response.content, 'html.parser')

print(soup.prettify())

<!DOCTYPE html>
<html lang="pt-br">
 <head id="Head1">
  <meta charset="utf-8"/>
  <meta content="width=device-width, initial-scale=1" name="viewport"/>
  <title>
   PROCESSO LEGISLATIVO ELETRÔNICO | CÂMARA MUNICIPAL DE CUIABÁ - MT
  </title>
  <!-- inicio: Favicon -->
  <link href="/spl/img/favicon/favicon-16x16.png?v=2" rel="icon" sizes="16x16" type="image/png"/>
  <link color="#1e9ee6" href="/spl/img/favicon/safari-pinned-tab.svg?v=2" rel="mask-icon"/>
  <meta content="#ffffff" name="msapplication-TileColor"/>
  <meta content="#ffffff" name="theme-color"/>
  <!-- fim: Favicon -->
  <!-- inicio: Social meta -->
  <meta content="website" property="og:type"/>
  <meta content="PROCESSO LEGISLATIVO ELETRÔNICO | CÂMARA MUNICIPAL DE CUIABÁ - MT" property="og:title"/>
  <meta content="Este Portal disponibiliza o Banco de Dados da Produção Legislativa de CÂMARA MUNICIPAL DE CUIABÁ - MT" property="og:description"/>
  <meta content="/spl/img/share.png?v=2" property="og:image"/>
  <meta content

In [3]:
soup


<!DOCTYPE html>

<html lang="pt-br">
<head id="Head1"><meta charset="utf-8"/><meta content="width=device-width, initial-scale=1" name="viewport"/><title>
	PROCESSO LEGISLATIVO ELETRÔNICO | CÂMARA MUNICIPAL DE CUIABÁ - MT
</title>
<!-- inicio: Favicon -->
<link href="/spl/img/favicon/favicon-16x16.png?v=2" rel="icon" sizes="16x16" type="image/png"/>
<link color="#1e9ee6" href="/spl/img/favicon/safari-pinned-tab.svg?v=2" rel="mask-icon"/>
<meta content="#ffffff" name="msapplication-TileColor"/><meta content="#ffffff" name="theme-color"/>
<!-- fim: Favicon -->
<!-- inicio: Social meta -->
<meta content="website" property="og:type"/>
<meta content="PROCESSO LEGISLATIVO ELETRÔNICO | CÂMARA MUNICIPAL DE CUIABÁ - MT" property="og:title"/>
<meta content="Este Portal disponibiliza o Banco de Dados da Produção Legislativa de CÂMARA MUNICIPAL DE CUIABÁ - MT" property="og:description"/>
<meta content="/spl/img/share.png?v=2" property="og:image"/>
<meta content="pt_BR" property="og:locale"><meta c

In [42]:
pagina = soup.find('div' , {'class': 'kt-widget5'}).get_text()

In [43]:
pagina

'\n\n\n\n\r\n                                            Projeto de Lei n° 270/2023\r\n                                        \n\n\r\n                                            DISPÕE SOBRE A APLICAÇÃO, NO ÂMBITO DA ADMINISTRAÇÃO DIRETA E INDIRETA DO MUNICÍPIO DE CUIABÁ, DA LEI FEDERAL Nº 13.019, DE 31 DE JULHO DE 2014, ALTERADA PELA LEI Nº 13.204, DE 14 DE DEZEMBRO DE 2015, QUE ESTABELECE O REGIME JURÍDICO DAS PARCERIAS COM ORGANIZAÇÕES DA SOCIEDADE CIVIL.\r\n                                        \n\nProcesso N°:\n36466/2023\nProtocolo N°:\n19660\nID:\n397172\nData:\n16/10/2023 10:28:15\nSituação:\nTramitando\n\n\nAutor(es) da Proposição:\nJohnny Everson\n\n\n\n\n\nACOMPANHE\n\n\nDETALHES\n\n\nPROCESSO\r\n                  \n\n\n\n\n\n\n\r\n                                            Projeto de Lei n° 269/2023\r\n                                        \n\n\r\n                                            DESMEMBRA O ORÇAMENTO NA LOA 2023 DA UNIDADE ORÇAMENTÁRIA 06.602 – FUNDO DE PR

In [6]:
import pandas as pd
import re

In [44]:
# split podemos passar o caracter que quero quebrar o texto:
pag_splt = pagina.strip().split('\n')

In [45]:
pag_splt

['Projeto de Lei n° 270/2023\r',
 '                                        ',
 '',
 '\r',
 '                                            DISPÕE SOBRE A APLICAÇÃO, NO ÂMBITO DA ADMINISTRAÇÃO DIRETA E INDIRETA DO MUNICÍPIO DE CUIABÁ, DA LEI FEDERAL Nº 13.019, DE 31 DE JULHO DE 2014, ALTERADA PELA LEI Nº 13.204, DE 14 DE DEZEMBRO DE 2015, QUE ESTABELECE O REGIME JURÍDICO DAS PARCERIAS COM ORGANIZAÇÕES DA SOCIEDADE CIVIL.\r',
 '                                        ',
 '',
 'Processo N°:',
 '36466/2023',
 'Protocolo N°:',
 '19660',
 'ID:',
 '397172',
 'Data:',
 '16/10/2023 10:28:15',
 'Situação:',
 'Tramitando',
 '',
 '',
 'Autor(es) da Proposição:',
 'Johnny Everson',
 '',
 '',
 '',
 '',
 '',
 'ACOMPANHE',
 '',
 '',
 'DETALHES',
 '',
 '',
 'PROCESSO\r',
 '                  ',
 '',
 '',
 '',
 '',
 '',
 '',
 '\r',
 '                                            Projeto de Lei n° 269/2023\r',
 '                                        ',
 '',
 '\r',
 '                                          

In [48]:
# Retirada das palavras em excesso retornas da Tag:
pag_splt_filtradas = [linha.strip() for linha in pag_splt if linha.strip() and linha.strip() not in ["ACOMPANHE", "DETALHES", "PROCESSO"]]

In [50]:
pag_splt_filtradas

['Projeto de Lei n° 270/2023',
 'DISPÕE SOBRE A APLICAÇÃO, NO ÂMBITO DA ADMINISTRAÇÃO DIRETA E INDIRETA DO MUNICÍPIO DE CUIABÁ, DA LEI FEDERAL Nº 13.019, DE 31 DE JULHO DE 2014, ALTERADA PELA LEI Nº 13.204, DE 14 DE DEZEMBRO DE 2015, QUE ESTABELECE O REGIME JURÍDICO DAS PARCERIAS COM ORGANIZAÇÕES DA SOCIEDADE CIVIL.',
 'Processo N°:',
 '36466/2023',
 'Protocolo N°:',
 '19660',
 'ID:',
 '397172',
 'Data:',
 '16/10/2023 10:28:15',
 'Situação:',
 'Tramitando',
 'Autor(es) da Proposição:',
 'Johnny Everson',
 'Projeto de Lei n° 269/2023',
 'DESMEMBRA O ORÇAMENTO NA LOA 2023 DA UNIDADE ORÇAMENTÁRIA 06.602 – FUNDO DE PREVIDÊNCIA SOCIAL DOS SERVIDORES PÚBLICOS DE CUIABÁ. (MENSAGEM Nº 33/2023)',
 'Processo N°:',
 '36391/2023',
 'Protocolo N°:',
 '19585',
 'ID:',
 '397045',
 'Data:',
 '11/10/2023 12:31:54',
 'Situação:',
 'Tramitando',
 'Autor(es) da Proposição:',
 'Executivo Municipal (Câmara Digital)',
 'Projeto de Lei n° 268/2023',
 'DISPÕE SOBRE A CONCESSÃO DE DIREITO REAL DE USO DO BEM PÚB

In [52]:
#import pandas as pd

# Sua string
#string_dados = pag_splt_filtradas

# Dividir a string em registros (linhas)
#registros = []
#registro_atual = []
#for linha in string_dados:
#    if linha.strip():  # Verifica se a linha não está em branco
#        registro_atual.append(linha.strip())
#    else:
#       if registro_atual:
#           registros.append(registro_atual)
#        registro_atual = []

# Se houver registros não adicionados no loop acima
#if registro_atual:
#   registros.append(registro_atual)

# Converter para DataFrame
#df = pd.DataFrame(registros, columns=["Projeto de Lei n° ", "Processo N°", "Protocolo N°,", "ID", "Data", "Situação", "Autor(es) da Proposição"])


In [47]:
df.head(20)

ACOMPANHE  \
0                          Projeto de Lei n° 270/2023   
1   DISPÕE SOBRE A APLICAÇÃO, NO ÂMBITO DA ADMINIS...   
2                                        Processo N°:   
3                            Autor(es) da Proposição:   
4                                           ACOMPANHE   
5                                            DETALHES   
6                                            PROCESSO   
7                          Projeto de Lei n° 269/2023   
8   DESMEMBRA O ORÇAMENTO NA LOA 2023 DA UNIDADE O...   
9                                        Processo N°:   
10                           Autor(es) da Proposição:   
11                                          ACOMPANHE   
12                                           DETALHES   
13                                           PROCESSO   
14                         Projeto de Lei n° 268/2023   
15  DISPÕE SOBRE A CONCESSÃO DE DIREITO REAL DE US...   
16                                       Processo N°:   
17                           Autor(es) da Proposição:   
18                                          ACOMPANHE   
19                                           DETALHES   

                                DETALHES       PROCESSO Projeto de Lei n°   \
0                                   None           None               None   
1                                   None           None               None   
2                             36466/2023  Protocolo N°:              19660   
3                         Johnny Everson           None               None   
4                                   None           None               None   
5                                   None           None               None   
6                                   None           None               None   
7                                   None           None               None   
8                                   None           None               None   
9                             36391/2023  Protocolo N°:              19585   
10  Executivo Municipal (Câmara Digital)           None               None   
11                                  None           None               None   
12                                  None           None               None   
13                                  None           None               None   
14                                  None           None               None   
15                                  None           None               None   
16                            36390/2023  Protocolo N°:              19584   
17  Executivo Municipal (Câmara Digital)           None               None   
18                                  None           None               None   
19                                  None           None               None   

   Processo N° Protocolo N°,     ID                 Data   Situação  \
0         None          None   None                 None       None   
1         None          None   None                 None       None   
2          ID:        397172  Data:  16/10/2023 10:28:15  Situação:   
3         None          None   None                 None       None   
4         None          None   None                 None       None   
5         None          None   None                 None       None   
6         None          None   None                 None       None   
7         None          None   None                 None       None   
8         None          None   None                 None       None   
9          ID:        397045  Data:  11/10/2023 12:31:54  Situação:   
10        None          None   None                 None       None   
11        None          None   None                 None       None   
12        None          None   None                 None       None   
13        None          None   None                 None       None   
14        None          None   None                 None       None   
15        None          None   None                 None       None   


In [11]:
pagina_limpa = re.sub(r'\s+', ' ', pagina)


In [12]:
pagina_limpa

' Projeto de Lei n° 270/2023 DISPÕE SOBRE A APLICAÇÃO, NO ÂMBITO DA ADMINISTRAÇÃO DIRETA E INDIRETA DO MUNICÍPIO DE CUIABÁ, DA LEI FEDERAL Nº 13.019, DE 31 DE JULHO DE 2014, ALTERADA PELA LEI Nº 13.204, DE 14 DE DEZEMBRO DE 2015, QUE ESTABELECE O REGIME JURÍDICO DAS PARCERIAS COM ORGANIZAÇÕES DA SOCIEDADE CIVIL. Processo N°: 36466/2023 Protocolo N°: 19660 ID: 397172 Data: 16/10/2023 10:28:15 Situação: Tramitando Autor(es) da Proposição: Johnny Everson ACOMPANHE DETALHES PROCESSO Projeto de Lei n° 269/2023 DESMEMBRA O ORÇAMENTO NA LOA 2023 DA UNIDADE ORÇAMENTÁRIA 06.602 – FUNDO DE PREVIDÊNCIA SOCIAL DOS SERVIDORES PÚBLICOS DE CUIABÁ. (MENSAGEM Nº 33/2023) Processo N°: 36391/2023 Protocolo N°: 19585 ID: 397045 Data: 11/10/2023 12:31:54 Situação: Tramitando Autor(es) da Proposição: Executivo Municipal (Câmara Digital) ACOMPANHE DETALHES PROCESSO Projeto de Lei n° 268/2023 DISPÕE SOBRE A CONCESSÃO DE DIREITO REAL DE USO DO BEM PÚBLICO MUNICIPAL E DÁ OUTRAS PROVIDÊNCIAS. (MENSAGEM Nº 34-202

In [ ]:
#Para teste:

import pandas as pd

# A string fornecida
string_dados = ['Projeto de Lei n° 270/2023\r',
 '                                        ',
 '',
 '\r',
 '                                            DISPÕE SOBRE A APLICAÇÃO, NO ÂMBITO DA ADMINISTRAÇÃO DIRETA E INDIRETA DO MUNICÍPIO DE CUIABÁ, DA LEI FEDERAL Nº 13.019, DE 31 DE JULHO DE 2014, ALTERADA PELA LEI Nº 13.204, DE 14 DE DEZEMBRO DE 2015, QUE ESTABELECE O REGIME JURÍDICO DAS PARCERIAS COM ORGANIZAÇÕES DA SOCIEDADE CIVIL.\r',
 '                                        ',
 '',
 'Processo N°:',
 '36466/2023',
 'Protocolo N°:',
 '19660',
 'ID:',
 '397172',
 'Data:',
 '16/10/2023 10:28:15',
 'Situação:',
 'Tramitando',
 '',
 '',
 'Autor(es) da Proposição:',
 'Johnny Everson',
 '',
 '',
 '',
 '',
 '',
 'ACOMPANHE',
 '',
 '',
 'DETALHES',
 '',
 '',
 'PROCESSO\r',
 '                  ',
 '',
 '',
 '',
 '',
 '',
 '',
 '\r']

Filtrar linhas vazias, remover linhas com "ACOMPANHE", "DETALHES" e "PROCESSO"
linhas_filtradas = [linha.strip() for linha in string_dados if linha.strip() and linha.strip() not in ["ACOMPANHE", "DETALHES", "PROCESSO"]]

Criar um DataFrame a partir das linhas filtradas
df = pd.DataFrame(linhas_filtradas, columns=["Dados"])

Limpar os dados removendo espaços e vírgulas
df['Dados'] = df['Dados'].str.replace(',', '').str.strip()

Separar os dados em duas colunas (Chave e Valor) usando a primeira ocorrência de ":"
df[['Chave', 'Valor']] = df['Dados'].str.split(':', 1, expand=True)

Resetar o índice do DataFrame
df.reset_index(drop=True, inplace=True)

Exibir o DataFrame
print(df)
